### PR TITLE
Pluggable ServiceBindingUsage

### PR DESCRIPTION
**Pull request by piotrmiskiewicz**
Confirm these statements before you submit your pull request:

- [x] I have read and submitted the required [Contributor Licence Agreements](https://github.com/kyma-project/community/blob/master/CONTRIBUTING.md#agreements-and-licenses).
- [x] This pull request follows the contributing guidelines, recommended Git workflow, and templates.
- [x] I have tested my changes.
- [x] I have updated the relevant documentation.
---

**Description**

Changes proposed in this pull request:

- Binding Usage Controller uses UsageKind resources as configuration, what resources can be used in 'usedBy' field in ServiceBindingUsage resoruce. The feature can be enabled by feature flag, which is disabled by default

**Issue link**

### CODE REVIEW

File: components/binding-usage-controller/examples/README.md [hunk 1]
```diff
@@ -115,3 +115,19 @@ To perform a clean-up, remove the Namespace:
 ```bash
 kubectl delete ns $namespace
 ```
+
+## Pluggable SBU
+
+The feature flag **APP_PLUGGABLE_SBU=true** change Binding Usage Controller logic. The following scenario shows, how to test it:
```
Comment (dpolitesap) [inline components/binding-usage-controller/examples/README.md]: The feature flag **APP_PLUGGABLE_SBU=true** changes Binding Usage Controller logic. The following scenario shows how to test it:

File: components/binding-usage-controller/examples/README.md [hunk 2]
```diff
@@ -115,3 +115,19 @@ To perform a clean-up, remove the Namespace:
...
+```bash
+APP_APPLIED_SBU_CONFIG_MAP_NAME=binding-usage-controller-process-sbu-spec APP_LOGGER_LEVEL=debug APP_PLUGGABLE_SBU=true APP_KUBECONFIG_PATH=~/.kube/config go run cmd/controller/main.go
+```
+
+2. Register **UsageKind** resources:
+```bash
+kubectl apply -f usage-kind/deployment.yaml
+```
+
+3. Follow the "Use the bindings on a Deployment" steps
```
Comment (dpolitesap) [inline components/binding-usage-controller/examples/README.md]: Follow the steps under "Use the bindings on a Deployment" steps

File: components/binding-usage-controller/internal/controller/usagekind/controller.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		Group:   kind.Spec.Resource.Group,
+		Kind:    kind.Spec.Resource.Kind,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "while creating resource interface provider")
+	}
+
+	supervisor := controller.NewGenericSupervisor(rip,
+		newLabelManipulator(kind.Spec.LabelsPath),
+		c.log.WithField("service", "generic-supervisor"))
```
Comment (aszecowka) [inline components/binding-usage-controller/internal/controller/usagekind/controller.go]: can we extend "generic-supervisor" with the name of resource?

File: components/binding-usage-controller/internal/controller/usagekind/controller.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+// Run begins watching and syncing.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	go func() {
+		<-stopCh
+		c.queue.ShutDown()
+	}()
+
+	c.log.Infof("Starting service binding usage controller")
```
Comment (aszecowka) [inline components/binding-usage-controller/internal/controller/usagekind/controller.go]: misleading log

File: components/binding-usage-controller/internal/controller/usagekind/controller.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	defer c.log.Infof("Shutting down service binding usage controller")
```
Comment (aszecowka) [inline components/binding-usage-controller/internal/controller/usagekind/controller.go]: misleading log

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	fixLabels := make(map[string]string)
+
+	// when
+	tracer.SetAnnotationAboutBindingUsage(obj, fixUsageName, fixLabels)
+
+	// then
+	got, err := tracer.GetInjectedLabels(obj, fixUsageName)
+	require.NoError(t, err)
+	assert.Len(t, got, len(fixLabels))
+	for k, v := range fixLabels {
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go]: you are right, I took the test code from sbu_tracer - we have the same problem there

File: components/binding-usage-controller/internal/controller/usagekind/controller.go (hunk 5 above)
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/usagekind/controller.go]: fixed

File: components/binding-usage-controller/internal/controller/usagekind/controller.go (hunk 4 above)
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/usagekind/controller.go]: fixed

File: components/binding-usage-controller/internal/controller/usagekind/controller.go (hunk 3 above)
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/usagekind/controller.go]: added logger field "kind"

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+// HasSynced returns true, because the GenericSupervisor does not use any informers/caches
+func (m *GenericSupervisor) HasSynced() bool {
+	return true
+}
+
+func (m *GenericSupervisor) executeUpdate(res *unstructured.Unstructured) error {
+	_, err := m.interfaceProvider.ResourceInterface(res.GetNamespace()).Update(res)
+	if err != nil {
+		return errors.Wrapf(err, "while updating %s %s", res.GetKind(), res.GetName())
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: added namespace

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+// GetInjectedLabels returns labels applied on given Deployment by usage controller
+func (m *GenericSupervisor) GetInjectedLabels(namespace, resourceName, usageName string) (map[string]string, error) {
+	res, err := m.interfaceProvider.ResourceInterface(namespace).Get(resourceName, metav1.GetOptions{})
+	switch {
+	case err == nil:
+	case apiErrors.IsNotFound(err):
+		return nil, NewNotFoundError(err.Error())
+	default:
+		return nil, errors.Wrap(err, "while listing Deployments")
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	m.labelSvc.EnsureLabelsAreDeleted(res, labels)
+
+	// remove annotations
+	err = m.tracer.DeleteAnnotationAboutBindingUsage(res, usageName)
+	if err != nil {
+		return errors.Wrap(err, "while deleting annotation tracing data")
+	}
+
+	if err := m.executeUpdate(res); err != nil {
+		return errors.Wrap(err, "while patching deployment")
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if err := m.executeUpdate(res); err != nil {
+		return errors.Wrap(err, "while updating deployment")
+	}
+
+	return nil
+}
+
+// EnsureLabelsDeleted ensures that given labels are deleted on Deployment
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	return &GenericSupervisor{
+		log: log,
+
+		tracer:            &genericUsageAnnotationTracer{},
+		interfaceProvider: interfaceProvider,
+		labelSvc:          labeler,
+	}
+}
+
+// EnsureLabelsCreated ensures that given labels are added to Deployment
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+	// apply new labels
+	m.labelSvc.EnsureLabelsAreApplied(res, labels)
+
+	if err := m.tracer.SetAnnotationAboutBindingUsage(res, usageName, labels); err != nil {
+		return errors.Wrap(err, "while setting annotation tracing data")
+	}
+
+	if err := m.executeUpdate(res); err != nil {
+		return errors.Wrap(err, "while updating deployment")
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+		interfaceProvider: interfaceProvider,
+		labelSvc:          labeler,
+	}
+}
+
+// EnsureLabelsCreated ensures that given labels are added to Deployment
+func (m *GenericSupervisor) EnsureLabelsCreated(namespace, resourceName, usageName string, labels map[string]string) error {
+	res, err := m.interfaceProvider.ResourceInterface(namespace).Get(resourceName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "while getting Deployment")
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: fixed

File: components/binding-usage-controller/internal/controller/generic_supervisor.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	// TODO (pluggable SBU cleanup): implement detecting labels conflict, see deployment_supervisor.go
+
+	// apply new labels
+	m.labelSvc.EnsureLabelsAreApplied(res, labels)
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_supervisor.go]: added error handling

File: components/binding-usage-controller/cmd/controller/main.go [hunk 15]
```diff
@@ -71,12 +77,25 @@ func main() {
...
+			cp,
+			log)
+
+	} else {
+		dSupervisor := controller.NewDeploymentSupervisor(k8sInformersFactory.Apps().V1beta2().Deployments(), k8sCli.AppsV1beta2(), log)
+		fnSupervisor := controller.NewKubelessFunctionSupervisor(kubelessInformerFactory.Kubeless().V1beta1().Functions(), kubelessCli.KubelessV1beta1(), log)
+		aggregator.Register(controller.KindDeployment, dSupervisor)
+		aggregator.Register(controller.KindKubelessFunction, fnSupervisor)
+	}
+	fatalOnError(err)
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/cmd/controller/main.go]: rearranged the code

File: components/binding-usage-controller/cmd/controller/main.go [hunk 16]
```diff
@@ -71,12 +77,25 @@ func main() {
...
+
+		kindController = usagekind.NewKindController(
+			bindingUsageInformerFactory.Servicecatalog().V1alpha1().UsageKinds(),
+			aggregator,
+			cp,
+			log)
+
+	} else {
+		dSupervisor := controller.NewDeploymentSupervisor(k8sInformersFactory.Apps().V1beta2().Deployments(), k8sCli.AppsV1beta2(), log)
+		fnSupervisor := controller.NewKubelessFunctionSupervisor(kubelessInformerFactory.Kubeless().V1beta1().Functions(), kubelessCli.KubelessV1beta1(), log)
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/cmd/controller/main.go]: kubeless factory moved into the else statement

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer.go [hunk 17]
```diff
@@ -0,0 +1,115 @@
+package controller
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+//go:generate mockery -name=usageBindingAnnotationTracer -output=automock -outpkg=automock -case=underscore
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer.go]: removed

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+type genericUsageBindingAnnotationTracer interface {
+	GetInjectedLabels(res *unstructured.Unstructured, usageName string) (map[string]string, error)
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer.go]: we will use it in the future

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+		gotValue, found := got[k]
+		assert.True(t, found)
+		assert.Equal(t, v, gotValue)
+	}
+
+	// when
+	tracer.DeleteAnnotationAboutBindingUsage(obj, fixUsageName)
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go]: checked

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer.go [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+	DeleteAnnotationAboutBindingUsage(res *unstructured.Unstructured, usageName string) error
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer.go]: discussed

File: components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/internal/controller/generic_sbu_tracer_test.go]: fixed

File: components/binding-usage-controller/examples/README.md (hunk 2 above)
Comment (piotrmiskiewicz) [reply components/binding-usage-controller/examples/README.md]: fixed

**Comment by aszecowka** [comment]
General problem:
I updated usedBy.name to point to not existing object and i got following error:
```
while ensuring proper labels on kind deployment: while checking if
        applied labels are equal with current ones: while getting injected labels:
        not found error: deployments.apps "redis-client-2" not found%!(EXTRA []interface
        {}=[])
```

File: resources/core/charts/service-catalog/charts/binding-usage-controller/templates/deployment-usage-kind.yaml [hunk 21]
```diff
@@ -0,0 +1,11 @@
+apiVersion: servicecatalog.kyma.cx/v1alpha1
```
[the remark this answers is not in the document]
Comment (piotrmiskiewicz) [reply resources/core/charts/service-catalog/charts/binding-usage-controller/templates/deployment-usage-kind.yaml]: RBAC updated

**Comment by piotrmiskiewicz** [comment]
regarding general problem about logging like:
"while ensuring proper labels on kind deployment: while checking if
        applied labels are equal with current ones: while getting injected labels:
        not found error: deployments.apps "redis-client-2" not found%!(EXTRA []interface
        {}=[])"

the corrupted message comes from K8s API

File: resources/core/charts/service-catalog/charts/binding-usage-controller/templates/cluster-role.yaml [hunk 22]
```diff
@@ -13,12 +13,12 @@ rules:
...
 - apiGroups: ["apps"]
   resources: ["deployments"]
-  verbs: ["get", "list", "watch", "patch"]
+  verbs: ["get", "list", "watch", "patch", "update"]
 - apiGroups: ["kubeless.io"]
   resources: ["functions"]
   verbs: ["get","list","watch", "update"]
 - apiGroups: ["servicecatalog.kyma.cx"]
-  resources: ["servicebindingusages"]
+  resources: ["servicebindingusages", "usagekinds"]
```
Comment (aszecowka) [inline resources/core/charts/service-catalog/charts/binding-usage-controller/templates/cluster-role.yaml]: now we are giving full access to operating on sbu and usagekinds, can we restrict it to read, watch and maybe list?

File: components/binding-usage-controller/README.md [hunk 23]
```diff
@@ -32,6 +32,7 @@ Use the following environment variables to configure the application:
 | **APP_PORT** | No | `3000` | The port on which the HTTP server listens. |
 | **APP_LOGGER_LEVEL** | No | `info` | Show detailed logs in the application. |
 | **APP_KUBECONFIG_PATH** | No |  | The path to the `kubeconfig` file that you need to run an application outside of the cluster. |
+| **APP_PLUGGABLE_SBU** | No | false | The feature flag, which enables binding usage pluggable by *UsageKind** resources.
```
Comment (dpolitesap) [inline components/binding-usage-controller/README.md]: The feature flag that enables pluggable binding usage by **UsageKind** resources.

File: components/binding-usage-controller/examples/README.md [hunk 24]
```diff
@@ -115,3 +115,19 @@ To perform a clean-up, remove the Namespace:
...
+```bash
+APP_APPLIED_SBU_CONFIG_MAP_NAME=binding-usage-controller-process-sbu-spec APP_LOGGER_LEVEL=debug APP_PLUGGABLE_SBU=true APP_KUBECONFIG_PATH=~/.kube/config go run cmd/controller/main.go
+```
+
+2. Register **UsageKind** resources:
+```bash
+kubectl apply -f usage-kind/deployment.yaml
+```
+
+3. Follow the steps under "Use the bindings on a Deployment" or "Use the bindings on a Function" steps
```
Comment (dpolitesap) [inline components/binding-usage-controller/examples/README.md]: Follow the steps under the steps **Use the bindings on a Deployment** or **Use the bindings on a Function**

**Pull request merged**